### DISCUSSION TITLE
Delegate maintenance to a few people

### DISCUSSION THREAD

**Discussion by crocket** [qutebrowser/qutebrowser#6704] [Ideas] [answered]
This project seems to be maintained(merged) by one man.

The maintainer is drowning due to lack of manpower.

He should delegate maintenance to a few people who will merge pull requests.

Linux maintenance is also delegated to developers.

**Comment by toofar** [comment]
The amount of responsibility placed on our diligent maintainer's shoulders is indeed unfortunate, but it's not a problem of permission. Several other people (including me) have permission to merge. Pressing the merge button is easy but it also implies a commitment to have the time and energy to maintain the changes you add to the project. That commitment is something I cannot give, I barely have the time to follow up on the issues that affect me and the features I care about, let alone the ones that other people care about. Given the activity patterns of other contributors I assume my experience is not unique.

For what it is worth, you are not alone in your concern and not engaging on constructive solutions is probably a bit negligent on my part. So I will do so. Not in a public forum though.

Lastly your request comes off as a bit demanding and low-effort. You might get more engagement if you try to frame your suggestions in a way that seems more encouraging, constructive and well thought out. (Linux is not really a comparable project in terms of resources for example.)

**Reply by crocket** [reply]
By the way, you can lower the standard for merging pull requests and deal with problems. This is going to accelerate learning.

You make a new release when things stabilize.

Pieter Hintjens created C4.1 contract and explained his standard for merging pull requests. C4.1 tells maintainers to merge something if it doesn't try to remove something from a stable API. It's far faster to break and fix than to do it perfectly the first time.

By just changing the standard, you can achieve a lot, I think. The problem is that maintainers are trying to be perfect. Perfectionism lowers the rate of learning and progress.

My observation is that perfectionism is a common problem among open-source project maintainers.

**Comment by The-Compiler** [comment, marked as answer]
I think @toofar has already said most of what I wanted to say. But still, I have some thoughts about this, so this is going to be long.

> This project seems to be maintained(merged) by one man.

Maintaining and merging are two wildly different things. Merging (as in pressing the merge button) code and getting code ready to merge are, again, two wildly different things.

Also, this statement isn't exactly true either. Like @toofar already pointed out, there are [5 other people](https://github.com/orgs/qutebrowser/people) with permissions to push code to the repository (2 of which set their membership to private).

The problem isn't having too few people with the permission to press the merge button. The problem is that there is a shortage of people willing to review code. Anyone can do so.

> The maintainer is drowning due to lack of manpower.

It's not exactly leading up to a constructive discussion when you're telling me how I'm apparently feeling. Care to leave your entitlement at the door perhaps?

> He should delegate maintenance to a few people who will merge pull requests.

Again, you come across as very entitled and demanding. I'll take some time to explain myself a bit below (mainly so I have something to link to for other people), but I'm not really interested in continuing this discussion if this is how you approach it.

Highly recommended reading: [The social contract of open source](https://snarky.ca/the-social-contract-of-open-source/). I don't owe you anything. In return, I also don't expect people to do more code review. It'd be nice if that was the case, but to some degree it's just a fact of life (or open source development) that people are generally more interested in writing new code than in reviewing it.

I'm seeing the same dynamics in many other successful open source projects. pytest, for example, has a core team of 18 people (some 5-10 currently active), yet also has a backlog of [42 open PRs](https://github.com/pytest-dev/pytest/pulls). Mypy [is at 97](https://github.com/python/mypy).

There [are some areas](https://github.com/qutebrowser/qutebrowser/blob/master/.github/CODEOWNERS) where other people are far more experienced with the area of the code and generally do reviews in those areas (@toofar, we should probably add you to that file for Greasemonkey things?).

But yes, it's the exception rather than the rule. It's a hard problem to solve. The solution isn't as easy as "just add more people!". The solution also isn't "just merge things faster!", because a messy/buggy codebase isn't exactly going to attract more contributors.

> By just changing the standard, you can achieve a lot, I think. The problem is that maintainers are trying to be perfect. Perfectionism lowers the rate of learning and progress.

Code review is hard, and I heavily disagree that "just press the merge button and deal with problems later" is the proper solution. Even more so for a project as security sensitive as qutebrowser. This is something you can do if you have the resources for it (i.e. people exclusively focused on fixing those issues fast), but it's not a suitable development model for a project in the size of qutebrowser.

Not "perfect", no. Yes, I acknowledge I can be pedantic in reviews sometimes, but it's not like I'm forcing people to do much more than the original problem they wanted to solve. I've very often merged pull requests improving the status quo but having some lose ends (which end up in a follow-up issues).

But whatever problem the contributor want to solve should be solved in a way it doesn't make future maintenance harder. I have sometimes ignored this and merged solutions for issues where I disagreed with the entire approach, because they fixed a short-time problem and seemed better than having no solution at all. Many of those things fell apart and caused headaches afterwards.

I'm not convinced at all that this approach is better in any way or form. "Making fast progress" isn't qutebrowser's goal. I'd rather have a bit of a slower pace of development instead of having to put out fires all the time. Because, let's face it, I'll be the one putting out those fires. I'm not willing to continue maintaining a code base over years if it's a frustrating code base to work with.

Like @toofar already mentioned, many casual contributors aren't really interested in actively maintaining their code in the future, and sometimes don't really care much about things like having automated tests, or things actively working in other environments (Qt versions, operating systems, etc.) than what they're running. They solved a problem they themselves are dealing with, and would like to upstream that work, but aren't really interested in the specifics on how it'll fit into the project.

Again, one of many realities of open source development which are probably hard to see until you're in the position of maintaining a somewhat popular open source project. I've certainly gained some new perspectives by maintaining qutebrowser for almost 8 years now. Another highly recommended reading, though a bit longer: [Nadia Eghbal: Working in Public: The Making and Maintenance of Open Source Software](https://press.stripe.com/#working-in-public). It explains many of those concepts and dynamics in a way I experienced during those years, but never quite could put into words.

Other people and projects might have different approaches. Just because they feel faster doesn't mean they are necessarily better. It's good to talk about those things, but please focus on having a constructive discussion and don't claim you know the perfect solutions for all those problems, because it's really not as easy as you claim.

I have some ideas in mind on how to improve things. The main idea still being an extension API, see #30. That way, people will be able to solve their problems without their solution having to be part of the qutebrowser core. Other people can use that solution, but then I'm not the one who is responsible when things go wrong there. Unfortunately, that extension API still isn't reality, a bit more about that below.

---

Now, let me explain a bit more about how the current PR backlog came to be, i.e. the more personal side of things. This is the "I've wanted to write this for a while, so I have something to link to" part.

If I had to put a timeline on things, I'd say things started getting out of hand somewhen in 2018. It was a very intense time in my studies and personal life, due to a variety of factors:

- I had a rather heavy exam session in my studies. Typically I had ignored contributions during the exam session and then picked things up again once exams were over. This was normal in the years before (I started my studies in 2016) and went well until then.
- In the semester after those exams, I wrote an important [term project](https://github.com/qutebrowser/qutebrowser-hsr-sa/releases/download/final/sa.pdf) for my studies, which ended up being on qutebrowser. This arrangement was great, but it meant that I was unable to handle any major contributions during that time. For one, because it would mudden the waters on what was my own work vs. contributed work during my term projects; but also because I was lacking time and already was doing a lot of work on qutebrowser things via the term project.
- At the same time, some bigger personal changes happened in my life. I broke up with my girlfriend of 7 years, I moved out from my mother's place, etc. etc. I didn't sleep very well for some time. Also add some health issues to the mix. I wouldn't say I burned out, but I certainly was close.
- I've tried to communicate as openly as possible that [I won't be able to look at PRs](https://github.com/qutebrowser/qutebrowser/commit/dd47e36444b3034faf7ec1d5fb7b4e446968c1b0) for some time. Yet, people generally ignored it. For some reason or another, actually, a lot *more* contributions than usual started coming in. Some people did help with reviews a bit, but generally, I was expected to pick up the lose ends.
- Due to all those factors, I kind of started dreading reviewing things and generally ignored them (sometimes successfully, sometimes less so) instead. It didn't help that some people are very demanding and entitled, without really having seen the whole picture.

Since then, I've been trying to pick up older PRs and getting them back into shape, while still taking care of keeping the whole thing running and not ignoring newer PRs. Recently (i.e. in the last month or two) this has slowed slowed down a bit again, more about that below.

In late 2019, I've finished my studies and, thanks to [donations](https://github.com/sponsors/The-Compiler/), am working part-time on qutebrowser. This has helped a lot with getting the backlog down and things into shape: Since I've started tracking PRs in a [project board](https://github.com/qutebrowser/qutebrowser/projects/4) in April 2020, 320 PRs have been merged (and [almost 1800 in total](https://github.com/qutebrowser/qutebrowser/pulls?q=is%3Apr+is%3Aclosed)), so put into perspective, things really aren't as bad as they might seem.

For new contributions where the author is actively interested in getting their code in, this will happen pretty quickly. For older contributions which require some additional work (from me, usually) to finish and update to the current codebase, things are taking a bit longer, but there is good progress there as well. Sometimes a contributor decides to pick up an older PR and get it ready again (#5987 comes to mind, but there have been others too). That also is a tremendous help and results in things being merged quickly.

Now, as for the older backlog mentioned above, I've had to find out that being employed 40% doesn't really mean I have 60% of my time available for qutebrowser. Due to the teaching I was involved in, I had weekly deadlines almost across the entire year, which took a lot of space in my mind even when I tried to focus on qutebrowser. I haven't really been able to deal with that as well as I liked.

This situation changed recently, however: I'm now teaching a new course I'm the main responsible person for (Automation with Python), to over 120 students. This week is the first week of the semester, thus, my workload was exceptionally high again for the last two months or so building everything up. This is why some more recent contributions piled up in the "Inbox" column of the [PR backlog](https://github.com/qutebrowser/qutebrowser/projects/4).

Once things for the course here are rolling smoothly, of course my workload will continue to be higher (teaching 120 students takes some effort, and I'm in on-site courses for two days a week). I still think I should be able to get back to the more recent/active PRs very soon.

As for the backlog, things will take a little longer - but the nice part of this arrangement is that I won't be teaching anything between January and August or so. Thus, I should be able to focus on some more bigger chunks of work again, such as older PRs and also the long-awaited extension API. How this will work out exactly in practice remains to be seen, of course, but I'm quite confident that it'll work better than the current arrangement did.

**Reply by crocket** [reply]
I thought at least being aware of a different approach can help even if you don't agree.

By considering different approaches, you can think less about fire fighting and more about big picture.

And, it seems that solving your money problem is going to help a lot. Trading and investment are one of ways to lift you above money problems. I'm also trying to learn trading. There are other ways, but it's the easiest way that I know.

I'm not saying you should learn trading, but being aware that there are creative ways to alleviate money problems can help.

Big picture thinking and thinking outside box can help.

**Reply by The-Compiler** [reply]
I honestly have no idea where you're getting a "money problem" from?! I'm in a quite comfortable situation financially, thank you very much.

If I had a money problem, I would have a full-time employment rather than only being employed 40%. Thanks to the qutebrowser community, I'm somewhere in the [top 13%](https://blog.tidelift.com/finding-2-most-maintainers-arent-paid-well-today-but-tidelift-is-having-an-impact) of people (world-wide) earning money via open-source donations. It's what allows me to work on qutebrowser as a part-time job rather than a hobby.

**Reply by crocket** [reply]
> Now, as for the older backlog mentioned above, I've had to find out that being employed 40% doesn't really mean I have 60% of my time available for qutebrowser. Due to the teaching I was involved in, I had weekly deadlines almost across the entire year, which took a lot of space in my mind even when I tried to focus on qutebrowser. I haven't really been able to deal with that as well as I liked.

It seems you want to work full-time on qutebrowser and still have some more free time left.

I know working part-time on qutebrowser is far better than nothing, but I sense your longing for more.

You may not like the way I write, but I'm actually rooting for qutebrowser.

**Reply by The-Compiler** [reply]
Jeez, let's just stop here. This is going nowhere.
